From 863db22affc61a7de6f8f2e0bd515d303bae9292 Mon Sep 17 00:00:00 2001
From: Christian Kohn <christian.kohn@xilinx.com>
Date: Fri, 26 Jul 2019 17:38:21 -0700
Subject: [PATCH] xlnxvideoscale: Add support for RGBx format

---
 sys/xlnxvideoscale/gstxlnxvideoscale.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/sys/xlnxvideoscale/gstxlnxvideoscale.c b/sys/xlnxvideoscale/gstxlnxvideoscale.c
index f380f20..4a293f8 100644
--- a/sys/xlnxvideoscale/gstxlnxvideoscale.c
+++ b/sys/xlnxvideoscale/gstxlnxvideoscale.c
@@ -98,7 +98,7 @@ struct xvpss_data
 
 #define XLNX_VIDEO_SCALE_CAPS \
     "video/x-raw, " \
-    "format = (string) {YUY2, UYVY, NV12, NV16, RGB, BGR, xRGB, GRAY8}, " \
+    "format = (string) {YUY2, UYVY, NV12, NV16, RGB, RGBx, BGR, xRGB, GRAY8}, " \
     "width = (int) [ 1, 3840 ], " \
     "height = (int) [ 1, 2160 ], " \
     "framerate = " GST_VIDEO_FPS_RANGE
@@ -187,6 +187,8 @@ get_xilinx_framebuf_format (GstVideoFormat gst_fourcc)
       return XILINX_FRMBUF_FMT_Y_UV8;
     case GST_VIDEO_FORMAT_RGB:
       return XILINX_FRMBUF_FMT_RGB8;
+    case GST_VIDEO_FORMAT_RGBx:
+      return XILINX_FRMBUF_FMT_RGBX8;
     case GST_VIDEO_FORMAT_BGR:
       return XILINX_FRMBUF_FMT_BGR8;
     case GST_VIDEO_FORMAT_xRGB:
@@ -203,6 +205,7 @@ get_xilinx_vpss_format (GstVideoFormat gst_fourcc)
 {
   switch (gst_fourcc) {
     case GST_VIDEO_FORMAT_RGB:
+    case GST_VIDEO_FORMAT_RGBx:
     case GST_VIDEO_FORMAT_BGR:
     case GST_VIDEO_FORMAT_xRGB:
       return XVIDC_CSF_RGB;
-- 
2.7.4

