From cc8291f9455f51a8070da2e62f8db71facd07483 Mon Sep 17 00:00:00 2001
From: vivekana <vivekana@xilinx.com>
Date: Wed, 30 Mar 2022 23:56:12 -0700
Subject: [PATCH] add mars driver to makefile

---
 drivers/media/i2c/Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index 0a5dcafa4193..6fcf98d6dd3f 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -134,4 +134,6 @@ obj-$(CONFIG_VIDEO_MAX9271_LIB)	+= max9271.o
 obj-$(CONFIG_VIDEO_RDACM20)	+= rdacm20.o
 obj-$(CONFIG_VIDEO_RDACM21)	+= rdacm21.o
 obj-$(CONFIG_VIDEO_ST_MIPID02) += st-mipid02.o
-obj-$(CONFIG_SDR_MAX2175) += max2175.o
+obj-$(CONFIG_SDR_MAX2175) += max2175.o  
+obj-$(CONFIG_VIDEO_MAX96705_LIB) += max96705.o
+obj-$(CONFIG_VIDEO_MARS) += mars.o
-- 
2.17.1

